runners/direct-java/src/main/java/org/apache/beam/runners/direct/WatermarkCallbackExecutor.java

```diff
@@ -116,14 +119,30 @@ public void callOnWindowExpiration(
    * Schedule all pending callbacks that must have produced output by the time of the provided
    * watermark.
    */
-  public void fireForWatermark(AppliedPTransform<?, ?, ?> step, Instant watermark) {
+  public void fireForWatermark(AppliedPTransform<?, ?, ?> step, Instant watermark)
+      throws InterruptedException {
     PriorityQueue<WatermarkCallback> callbackQueue = callbacks.get(step);
     if (callbackQueue == null) {
       return;
     }
     synchronized (callbackQueue) {
+      List<Runnable> toFire = new ArrayList<>();
       while (!callbackQueue.isEmpty() && callbackQueue.peek().shouldFire(watermark)) {
-        executor.execute(callbackQueue.poll().getCallback());
+        toFire.add(callbackQueue.poll().getCallback());
+      }
+      if (!toFire.isEmpty()) {
+        CountDownLatch latch = new CountDownLatch(toFire.size());
```

Comment:
Instead of managing the latch yourself, you could gather a list of futures and wait on it. It would be a bit higher level programming style, and not depend on capturing the latch in the closure, etc.

Reply:
If I understand it correctly, then this in unfortunately currently not possible, because the `executor` is not `ExecutorService`, but simply `Executor`, which returns `void`, instead of `Future`. I cannot simply change it to `ExecutorService`, because `EvaluationContext` passes there `MoreExecutors.directExecutor()`, which is not `ExecutorService`.

Reply:
Hm, my previous comment might suggest, that the change is not needed at all. And maybe it is true. :-) I added this code as part of my "I'm really desperate" period while solving the issue. Just for being sure, that this doesn't break in the future, I'd like to keep it as is, if it is okay.

Reply:
Fair enough :-). Since it passes and solves a problem, let's definitely merge.